File: OpenRaccoon/Sources/RaccoonKit/Models/BridgeConnection.swift
import Foundation

public struct BridgeConnection: Codable, Identifiable, Sendable, Equatable {
    public let id: String
    public let userID: String
    public var platform: Platform
    public var method: BridgeMethod
    public var status: BridgeStatus
    public var metadata: [String: AnyCodable]?
    public var lastSyncAt: Date?
    public let createdAt: Date
    public var updatedAt: Date

    public enum Platform: String, Codable, Sendable {
        case telegram
        case whatsapp
        case signal
        case discord
    }

    public enum BridgeMethod: String, Codable, Sendable {
        case userLevel = "user_level"
        case bot
        case cloudAPI = "cloud_api"
    }

    public enum BridgeStatus: String, Codable, Sendable {
        case connected
        case reconnecting
        case disconnected
        case error
    }

    enum CodingKeys: String, CodingKey {
        case id
        case userID = "userId"
        case platform
        case method
        case status
        case metadata
        case lastSyncAt
        case createdAt
        case updatedAt
    }

    public init(
        id: String,
        userID: String,
        platform: Platform,
        method: BridgeMethod,
        status: BridgeStatus = .disconnected,
        metadata: [String: AnyCodable]? = nil,
        lastSyncAt: Date? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.userID = userID
        self.platform = platform
        self.method = method
        self.status = status
        self.metadata = metadata
        self.lastSyncAt = lastSyncAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

File: open_raccoon/apps/raccoon_bridges/lib/raccoon_bridges/bridge_connection.ex
defmodule RaccoonBridges.BridgeConnection do
  @moduledoc """
  Bridge connection schema for external platform integrations.
  """
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "bridge_connections" do
    belongs_to(:user, RaccoonAccounts.User)

    field(:platform, Ecto.Enum, values: [:telegram, :whatsapp, :signal, :discord])
    field(:method, Ecto.Enum, values: [:user_level, :bot, :cloud_api])

    field(:status, Ecto.Enum,
      values: [:connected, :reconnecting, :disconnected, :error],
      default: :disconnected
    )

    field(:encrypted_credentials, :binary)
    field(:metadata, :map, default: %{})
    field(:last_sync_at, :utc_datetime_usec)

    has_many(:conversations, RaccoonChat.Conversation, foreign_key: :bridge_id)

    timestamps(type: :utc_datetime_usec)
  end

  def changeset(bridge, attrs) do
    bridge
    |> cast(attrs, [
      :user_id,
      :platform,
      :method,
      :status,
      :encrypted_credentials,
      :metadata,
      :last_sync_at
    ])
    |> validate_required([:user_id, :platform, :method])
    |> validate_inclusion(:platform, [:telegram, :whatsapp, :signal, :discord])
    |> validate_inclusion(:method, [:user_level, :bot, :cloud_api])
    |> validate_inclusion(:status, [:connected, :reconnecting, :disconnected, :error])
    |> unique_constraint([:user_id, :platform, :method])
    |> foreign_key_constraint(:user_id)
  end
end
